---
title: "HW1"
author: "Anastasiia Gurianova"
date: "2024-03-15"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 0. Kallisto
### 0.1 Kallisto installation
```{bash kallisto_install, eval = FALSE}
# install to local machine

git clone https://github.com/pachterlab/kallisto
cd kallisto
mkdir build
cd build

# install cmake, hdf5

cmake  .. -DUSE_HDF5=ON
make

# check if all is ok

kallisto

# if not add kallisto directory to PATH
```

### 0.2 Kallisto index

kallisto index -i reference.idx reference.fasta

referense transcriptome downloaded from 
https://ftp.ensembl.org/pub/release-111/fasta/mus_musculus/cdna/
```{bash kallisto_index, eval = FALSE}
kallisto index -i reference.idx /Users/lune/Downloads/Mus_musculus.GRCm39.cdna.all.fa.gz
```

### 0.3 Kallisto quant

kallisto quant -i reference.idx -o results --single -l 200 -s 30 downloaded.fasta

-l 200 -s 30 for single end mode only
```{bash kallisto_quant, eval = FALSE}
kallisto quant -i reference.idx -o SRR7425017_GSM3215429_Foam_intimal_macrophages_rep_1_Bulk_RNA-seq_Mus_musculus_RNA-Seq --single -l 200 -s 30 /Users/lune/Education/Transcriptomics/SRR7425017_GSM3215429_Foam_intimal_macrophages_rep_1_Bulk_RNA-seq_Mus_musculus_RNA-Seq.fastq.gz
```

## 1. Download data
Here I load kallisto results (abundance.h5 files) and phenotable (prepared it in accordance with kallisto results and samples annotation on GEO)
```{r Download data, warning=FALSE, message=FALSE}
library(readr)
library(tximport)
library(rhdf5)

samples <- read_tsv("/Users/lune/Education/Transcriptomics/HW1/phenotable.tsv")
files <- file.path(samples$path, "abundance.h5") 
setwd("/Users/lune/Education/Transcriptomics/HW1/kallisto_results/")
txi <- tximport(files, type = 'kallisto', txOut = T)
```


## 2. Deseq
### 2.1 DE transcripts
```{r Deseq, warning=FALSE, message=FALSE}
library(tidyverse)
library(DESeq2)

# preparation
ddsTxi <- DESeqDataSetFromTximport(txi,
                                   colData = samples,
                                   design = ~ condition)
dds <- ddsTxi[rowSums(counts(ddsTxi)) >= 10,] # remove low expewssed genes
dds$condition <- relevel(dds$condition, ref = "Non_foam") # relevel condition

# DESeq
dds <- DESeq(dds)
vsd <- varianceStabilizingTransformation(dds)
res <- results(dds, name = 'condition_Foam_vs_Non_foam')
res <- res %>%
  as.data.frame %>% 
  rownames_to_column("transcript") 
head(res)

# additional info and functions
#res005 <- results(dds, alpha=0.05)
#resultsNames(dds)
#model.matrix(~0+condition, samples) %>% View()
```

### 2.2 DE genes
Here I add two additional column to res table because we have only transcript names initially. But gene names are crucial for the downstream analysis. I add columns with gene names in ENSEMBL and ENTREZ format. To find annotation I use biomaRt.
```{r Deseq_Gene_names, warning=FALSE, message=FALSE}
library(biomaRt)
library(clusterProfiler)

# transcript_annotation with biomaRT
mart <- useMart(biomart = "ensembl", dataset = "mmusculus_gene_ensembl")

httr::set_config(httr::config(timeout = 60)) # cos I had a problem with connection
transcript_annotation = getBM(attributes = c('ensembl_gene_id','ensembl_transcript_id_version','entrezgene_id'), 
                              filters = 'ensembl_transcript_id_version', 
                              values = res$transcript, 
                              mart = mart)

# add column with gene names in ENTREZ and ENSEMBL format into res
res$gene_ENTREZ = transcript_annotation$entrezgene_id[match(res$transcript, transcript_annotation$ensembl_transcript_id_version)]
res$gene_ENSEMBL = transcript_annotation$ensembl_gene_id[match(res$transcript, transcript_annotation$ensembl_transcript_id_version)]

# and update column order
res = res[c(8:9,1:7)]
head(res) # gene_ENTREZ is NA at the start of the table
```

### 2.3 Significant results
threshold padj: < 0.05, threshold log2FC: > or < 0
```{r Deseq_sign_res, warning=FALSE, message=FALSE}
sign_up <- res %>% filter(log2FoldChange > 0) %>% filter(padj < .05)
print(paste("Number of up-regulated genes:", dim(sign_up)[1]))

sign_dw <- res %>% filter(log2FoldChange < 0) %>% filter(padj < .05)
print(paste("Number of down-regulated genes:", dim(sign_dw)[1]))
```


## 3. Consistency tests
### 3.1 Volcano plot
```{r Consistency_volcano, warning=FALSE, message=FALSE}
res %>%
  ggplot(aes(log2FoldChange, -log10(padj), color = padj < 0.05))+
  geom_point()+
  scale_color_manual(values=c("black", "red"))+
  xlim(c(-2.5, 2.5))
```

### 3.2 PCA
```{r Consistency_PCA, warning=FALSE, message=FALSE}
plotPCA(vsd, intgroup=c("condition"))
```

### 3.3 Single gene expression
```{r Consistency_single_gene_expression, warning=FALSE, message=FALSE}
plotCounts(dds, gene="ENSMUST00000222971.2", intgroup="condition")
```

### 3.4 Heatmap
Heatmap of 30 most expressed genes
```{r Consistency_heatmap, warning=FALSE, message=FALSE}
library(pheatmap)

select <- order(rowMeans(counts(dds,normalized=TRUE)),
                decreasing=TRUE)[1:30]
df <- as.data.frame(colData(dds)[,c("condition")])
colnames(df) <- "condition"
pheatmap(assay(vsd)[select,], cluster_rows=T, show_rownames=T,
         cluster_cols=T, annotation_col=df)
```

### 3.5 Distance matrix
```{r Consistency_distance_matrix, warning=FALSE, message=FALSE}
library(pheatmap)
library(RColorBrewer)

sampleDists <- dist(t(assay(vsd)))

sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- vsd$condition
colnames(sampleDistMatrix) <- vsd$condition
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")))(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         color = colors)
```

## 4.1 KEGG enrichment
### 4.1.1 Downregulated genes
```{r kegg_down, warning=FALSE, message=FALSE}
library(clusterProfiler)

KEGG_enrich <- enrichKEGG(sign_dw$gene_ENTREZ, organism = "mmu",keyType =  'ncbi-geneid')

# Visualize
barplot(KEGG_enrich)
dotplot(KEGG_enrich)
cnetplot(KEGG_enrich)
```

### 4.1.2 Upregulated genes
```{r kegg_up, warning=FALSE, message=FALSE}
library(clusterProfiler)

KEGG_enrich <- enrichKEGG(sign_up$gene_ENTREZ, organism = "mmu",keyType =  'ncbi-geneid')

# Visualize
barplot(KEGG_enrich)
dotplot(KEGG_enrich)
cnetplot(KEGG_enrich)
```

## 4.2 GO enrichment
### 4.2.1 Downregulated genes
```{r go_down, warning=FALSE, message=FALSE}
library(org.Mm.eg.db)
library(clusterProfiler)

GO_enrich <- enrichGO(sign_dw$gene_ENTREZ, "org.Mm.eg.db", keyType = "ENTREZID", ont = "MF")
head(GO_enrich@result)

#dotplot
dotplot(GO_enrich, showCategory = 20)

#emapplot
GO_enrich <- enrichplot::pairwise_termsim(GO_enrich)
emapplot(GO_enrich)

#goplot
goplot(GO_enrich)
```

### 4.2.1 Upregulated genes
```{r go_up, warning=FALSE, message=FALSE}
library(org.Mm.eg.db)
library(clusterProfiler)

GO_enrich <- enrichGO(sign_up$gene_ENTREZ, "org.Mm.eg.db", keyType = "ENTREZID", ont = "MF")
head(GO_enrich@result)

#dotplot
dotplot(GO_enrich, showCategory = 20)

#emapplot
GO_enrich <- enrichplot::pairwise_termsim(GO_enrich)
emapplot(GO_enrich)

#goplot
goplot(GO_enrich)
```

## 4.3 GSEA
### 4.3.1 ranks
```{r GSEA_ranks, warning=FALSE, message=FALSE}
translated_names <- res %>% 
  na.omit() %>%
  pull(gene_ENSEMBL) %>% 
  bitr("ENSEMBL", c("GENENAME", "ENTREZID"), "org.Mm.eg.db") %>% 
  distinct(ENSEMBL, .keep_all = T)

ranks_for_gsea <- res %>% 
  na.omit() %>% 
  arrange(desc(stat)) %>% 
  mutate(ENSEMBL = gene_ENSEMBL) %>% 
  left_join(translated_names) %>% 
  mutate(NAME = ifelse(is.na(ENTREZID), ENSEMBL, ENTREZID)) %>% 
  transmute(NAME, stat) %>% 
  deframe()

head(ranks_for_gsea)
```

### 4.3.2 pathways
```{r GSEA_pathways, warning=FALSE, message=FALSE}
#https://data.wikipathways.org/current/gmt/
pathways <- fgsea::gmtPathways("/Users/lune/Education/Transcriptomics/HW1/wikipathways-20240310-gmt-Mus_musculus.gmt")
```

### 4.3.3 GSEA
```{r GSEA_results, warning=FALSE, message=FALSE}
fgsea_results <- fgsea::fgseaMultilevel(pathways, ranks_for_gsea)
head(fgsea_results[order(pval),])
```

### 4.3.4 GSEA plot for one pathway
```{r GSEA_plot, warning=FALSE, message=FALSE}
fgsea::plotEnrichment(pathways[["Electron transport chain%WikiPathways_20240310%WP295%Mus musculus"]], ranks_for_gsea) + labs(title="Electron transport chain")
```

### 4.3.5 GSEA top and bottom 10 pathways
```{r GSEA_10, warning=FALSE, message=FALSE}
topPathwaysUp <- fgsea_results[ES > 0][head(order(pval), n=10), pathway]
topPathwaysDown <- fgsea_results[ES < 0][head(order(pval), n=10), pathway]
topPathways <- c(topPathwaysUp, rev(topPathwaysDown))
topPathways = gsub("%WikiPathways_20240310.*","",topPathways) #for plot
names(pathways) = gsub("%WikiPathways_20240310.*","",names(pathways)) #for plot
fgsea_results$pathway = gsub("%WikiPathways_20240310.*","",fgsea_results$pathway) # for plot
fgsea::plotGseaTable(pathways[topPathways], ranks_for_gsea, fgsea_results,
                     gseaParam=0.5)
```

### 4.3.6 GSEA with clusterProfiler
```{r GSEA_clusterProfiler, warning=FALSE, message=FALSE}
#library(ggridges)
#library(org.Mm.eg.db)
#library(clusterProfiler)

GO_gsea = gseGO(ranks_for_gsea, ont = "ALL", 'org.Mm.eg.db', eps = 0)
ridgeplot(GO_gsea)
#gseaplot(GO_gsea, 1, title = GO_gsea@result$Description[[1]]))
```
